add Sec-CH-Prefers-Reduced-Motion content

chrisdavidmills opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

This PR adds content for the `Sec-CH-Prefers-Reduced-Motion` header; see my research document for more details: https://docs.google.com/document/d/1ENCgZ66Jsf40iMK1yatqQLeEHu9NC07Fhd6tKlgj_Bs/edit

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

Readers will want to know about this new HTTP header

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

chrisdavidmills committed:
add Sec-CH-Prefers-Reduced-Motion content

tomayac commented on files/en-us/web/http/headers/index.md:
```diff
@@ -84,6 +84,8 @@ The different categories of client hints are listed below.
 
 The [UA client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) are request headers that provide information about the user agent and the platform/architecture on which it is running:
```
```suggestion
The [UA client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) are request headers that provide information about the user agent and the platform/architecture it is running on, and about user preferences:
```

tomayac commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -0,0 +1,90 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}
+
+The **`Sec-CH-Prefers-Reduced-Motion`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header provides the user agent's reduced motion preference setting.
```
```suggestion
The **`Sec-CH-Prefers-Reduced-Motion`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header provides the user agent's reduced motion preference, typically (but not necessarily) reflecting the operating system's corresponding setting.
```

tomayac commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -0,0 +1,90 @@
+
+The **`Sec-CH-Prefers-Reduced-Motion`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header provides the user agent's reduced motion preference setting.
+
+If a server declares to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Reduced-Motion`, the client can then respond with this header to indicate the user's preference for reduced motion. The server can send the client appropriate CSS to reduce the motion of any animations presented on subsequent rendered content.
```
```suggestion
If a server declares to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Reduced-Motion`, the client then responds with this header to indicate the user's preference for reduced motion. The server can send the client appropriately adapted content, for example, CSS to reduce the motion of any animations presented on subsequent rendered content.
```

tomayac commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
````diff
@@ -0,0 +1,90 @@
+Accept-CH: Sec-CH-Prefers-Reduced-Motion
+```
+
+The client then retries the request, telling the server via `Sec-CH-Prefers-Reduced-Motion` that it has a user preference for reduced-motion animations.
````
```suggestion
The client in future requests then tells the server via `Sec-CH-Prefers-Reduced-Motion` that it has a user preference for reduced-motion animations.
```

The retry flow only happens if the header is marked as a critical client hint; see https://web.dev/user-preference-media-features-headers/#how-it-works for a full example. You might want to note this aspect.

hamishwillee replied:
This is still confusing. Probably because it is different than other cases - or at least how this was about 9 months ago (last time I worked on a client hint).

I _think_ there are now three options for hints
1. "Normal": Server set Accept-CH for headers it works with. Client responds with the Sec-CH headers it is interested in sending them for all future requests. These are sent until end of browser session unless new set of Accept-CH is sent
2. "Low entropy hints": sent by User agent all the time if they are interested in them, without Accept-CH.
3. Critical hints: Monitor Critical-CH and Accept-CH in response. If these don't match the values you sent for your Sec-CH in the last request then "retry".
   - I assume "retry" means "throw away the response" and send the request again as though it were the first request?
   - I assume that IF there is a match then you can continue to just resend the same as a normal hint in subsequent requests (or at least that is what the spec infers).

So what you say now is:

> The client then retries the request, telling the server via `Sec-CH-Prefers-Reduced-Motion` that it has a user preference for reduced-motion animations.

1. But what you have marked as critical is "`Critical-CH: Sec-CH-Prefers-Color-Scheme`" - i.e. the color scheme, not the reduced preference. My interpretation of the spec is that you WILL retry the request because you didn't send Sec-CH-Prefers-Color-Scheme in the initial request. This is confusing because you'll be retrying it for the `Sec-CH-Prefers-Color-Scheme`, and only incidentally sending the updated `Sec-CH-Prefers-Reduced-Motion`. Make sense? It's not wrong - it's just that you really don't need to have the retry to demonstrate Sec-CH-Prefers-Reduced-Motion unless it is critical.
2. It isn't clear which of these need to be critical.
3. The spec says that whatever appears in Critical-CH MUST appear in Accept-CH. So there is a header mismatch
4. IF this hint does not always have to be critical (i.e. there is no "critical header hint set" in the same way as there are low entropy hints) there is a another flow where the message is sent as normal. I think we need to consider both cases. For example, you do your retry here then I would expect some info about subsequent requests including the same header until Xxxxx (where Xxxx is probably the set of Critical-CH changing).

hamishwillee replied:
PS I might do this as two examples. One with Sec-CH-Prefers-Reduced-Motion as the critical header (swapping them over), and one with no critical headers. Depends a bit on how the set of critical headers is defined, and whether `Sec-CH-Prefers-Reduced-Motion` is likely to always be one.

chrisdavidmills replied:
See updated text. I think this is less confusing now. `Sec-CH-Prefers-Color-Scheme` was a typo, and I've now updated it to `Sec-CH-Prefers-Reduced-Motion`

hamishwillee replied:
Yes it is. Thanks.

tomayac reviewed (commented):
Thanks for documenting this! Some comments with clarifications and suggestions. Once applied, you can then take this content and flip the content for `sec-ch-prefers-color-scheme`.

chrisdavidmills committed:
Update files/en-us/web/http/headers/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
making a few tweaks to tomayacs updates

chrisdavidmills commented:
Thanks @tomayac ! I've made a few small tweaks to your wording, and I also added in the `Critical-CH` header to the example, as it seems useful. Now we have to document that header too, but I'm hopefully going to be able to leave it to a future update.

hamishwillee commented:
HI @chrisdavidmills 

This probably also needs to be added to https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints - possibly in a new class of hint.

Note, that this doc is the MDN equivalent to https://web.dev/user-agent-client-hints/ . If you look at the two side by side you can see that the MDN version, while more up to date in terms of the headers covered is still missing some key bits of information
- cross-origin usage
- A could of real examples
- More clear highlighting of the interaction with the JavaScript API.

One day that might get to my list, but feel free to jump on it :-)

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -0,0 +1,91 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}
+
+The **`Sec-CH-Prefers-Reduced-Motion`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header provides the user agent's reduced motion preference, typically (but not necessarily) reflecting the operating system's corresponding setting.
```
I like docs such that you only need to read the first sentence to know whether the topic is relevant.
- Can we clarify that this is related specifically to animations - CSS and JavaScript. 
- It would also be great (if possible) if we could characterise motion - e.g. are we talking speed, or amplitude, or just "how frenetic/lazy it is"?
-  Bonus points if you can explain why we care about this - I assume it is for accessibility? Ideally in first paragraph, but OK in description somewhere.

Possibly this - defining reduced motion in terms of the affects it has for accessibility. 

```suggestion
The **`Sec-CH-Prefers-Reduced-Motion`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header indicates the user agent's preference for animations to be displayed with a reduced motion.

This is intended to allow the server to provide accessible content for users affected by rapid motion, including those for which rapid visual effects may trigger seizures.
Typically (but not necessarily) this preference reflects a correpsonding setting in the  operating system.
```

chrisdavidmills replied:
OK, added clarifications in my next commit; I don't want to say too much, as there are more details available on the prefers-reduced-motion page that is linked, but yeah, good to clarify.

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
````diff
@@ -0,0 +1,91 @@
+Host: example.com
+```
+
+The server responds, telling the client via {{httpheader("Accept-CH")}} and {{httpheader("Critical-CH")}} that it accepts `Sec-CH-Prefers-Reduced-Motion` and considers it a critical client hint:
````
Critical client hint is a new one on me. Can we add a definition of that to "Client_hints" doc? If not, consider omitting it here - it isn't "necessary" to the example.

chrisdavidmills replied:
I have updated the text to clarify that `Critical-CH` is not mandatory, and that specifying it is what caused the browser to automatically retry to request with `Sec-CH-Prefers-Reduced-Motion`. I've also added a small section about it in the Client_hints doc.

chrisdavidmills commented:
Cool, thanks for the review @hamishwillee . To begin with, I've added a line about this specific classic of UA hints to the "User-agent client hints" section of the Client hints page, as they are definitely in that category. That'll be in the next commit.

chrisdavidmills committed:
making fixes for hamishwillee comments

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
````diff
@@ -0,0 +1,91 @@
+Host: example.com
+```
+
+The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Reduced-Motion`. {{httpheader("Critical-CH")}} is also used in this case to specify that `Sec-CH-Prefers-Reduced-Motion` is considered a critical client hint:
````
Note, I link to critical client hint docs - but I have used a different anchor URL - proposing to change that too:

```suggestion
The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Reduced-Motion`. In this example {{httpheader("Critical-CH")}} is also used, indicating that `Sec-CH-Prefers-Reduced-Motion` is considered a [critical client hint](/en-US/docs/Web/HTTP/Client_hints#critical_client_hints):
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
````diff
@@ -0,0 +1,91 @@
+Host: example.com
+Sec-CH-Prefers-Reduced-Motion: "reduce"
+```
+
````
What happens in subsequent requests? Something like 

```suggestion

The client will include the header in subsequent requests in the current session, unless the `Accept-CH` changes in responses to indicate that it is no longer supported by the server.

```

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -72,13 +72,36 @@ The high entropy hints are those that have the potential to give away more infor
 The decision might be based on user preferences, a permission request, or the permission policy.
 All client hints that are not low entropy hints are considered high entropy hints.
 
+## Specifying critical client hints
```
```suggestion
## Critical client hints
```

hamishwillee committed:
Move User preference media features client hints to own section

hamishwillee committed:
Update files/en-us/web/http/client_hints/index.md

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -86,6 +107,11 @@ Client hints are available to web page JavaScript via the [User Agent Client Hin
 > UA client hints provide a more efficient and privacy preserving way of getting the desired information.
 > They are eventually expected to replace this older approach.
 
+### User preference media features client hints
```
FYI, moved this into its own section below. This is a useful way of presenting the sets as they come along.

hamishwillee replied:
> To begin with, I've added a line about this specific classic of UA hints to the "User-agent client hints" section of the Client hints page, as they are definitely in that category. That'll be in the next commit.

PS Reasoning is that yes, they are user agent hints, but they are quite specific in the way they are set and how someone might look for them (i.e. when you compare them to the other hints in that category). So this is a useful grouping in the same ways as device hints are.

chrisdavidmills replied:
Yup, this looks good.

hamishwillee committed:
Update files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -72,6 +72,27 @@ The high entropy hints are those that have the potential to give away more infor
 The decision might be based on user preferences, a permission request, or the permission policy.
 All client hints that are not low entropy hints are considered high entropy hints.
 
+## Critical client hints
+
```
```suggestion
## Critical client hints

A _critical client hint_ is one where applying the response may significantly change the rendered page, potentially in a way that is jarring or will significantly affect usability, and which must therefore be applied before the content is rendered.
For example, `Sec-CH-Prefers-Reduced-Motion` is commonly treated as a critical hint, because it might markedly affect the behaviour of animations, and because a user who chooses this preference really needs for it to be set.

```

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -72,6 +72,27 @@ The high entropy hints are those that have the potential to give away more infor
 
+## Critical client hints
+
+You can use the {{HTTPHeader("Critical-CH")}} response header along with `Accept-CH` to specify that an accepted client hint is also a _critical client hint_. User agents should automatically send the server that client hint to indicate their preference in that regard.
```
FWIW "
- You" better as "server" or "user agent" when documenting HTTP. Otherwise the flow can be confusing.
- "User agents should automatically send the server that client hint to indicate their preference in that regard." is confusing/too early. 

```suggestion
A server can use the {{HTTPHeader("Critical-CH")}} response header along with `Accept-CH` to specify that an accepted client hint is also a critical client hint (note that a header in `Critical-CH` must also appear in `Accept-CH`).
User agents receiving a response with `Critical-CH` must check if the indicated critical headers were sent in the original request. If not, then the user agent will retry the request rather than rendering the page.
This approach ensures that client preferences set using critical client hints are always used, even if not included in the first request, or if the server configuration changes.

```

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -72,6 +72,27 @@ The high entropy hints are those that have the potential to give away more infor
+
+You can use the {{HTTPHeader("Critical-CH")}} response header along with `Accept-CH` to specify that an accepted client hint is also a _critical client hint_. User agents should automatically send the server that client hint to indicate their preference in that regard.
+
+For example, in this case, the server tells a client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Reduced-Motion`. {{httpheader("Critical-CH")}} is also used to specify that `Sec-CH-Prefers-Reduced-Motion` is considered a critical client hint:
```
```suggestion
For example, in this case, the server tells a client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Reduced-Motion`, and {{httpheader("Critical-CH")}} is used to specify that `Sec-CH-Prefers-Reduced-Motion` is considered a critical client hint:
```

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
````diff
@@ -72,6 +72,27 @@ The high entropy hints are those that have the potential to give away more infor
+Critical-CH: Sec-CH-Prefers-Reduced-Motion
+```
+
+The client automatically retries the request, telling the server via `Sec-CH-Prefers-Reduced-Motion` that it has a user preference for reduced-motion animations.
````
```suggestion
As `Sec-CH-Prefers-Reduced-Motion` is a critical hint that was not in the original request, the client automatically retries the request — this time telling the server via `Sec-CH-Prefers-Reduced-Motion` that it has a user preference for reduced-motion animations.
```

hamishwillee committed:
Additional detail in critical client hints

hamishwillee commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -72,6 +72,33 @@ The high entropy hints are those that have the potential to give away more infor
 The decision might be based on user preferences, a permission request, or the permission policy.
 All client hints that are not low entropy hints are considered high entropy hints.
 
+## Critical client hints
```
FYI, I have rewritten this to explain how client hints work. We could go further with all of this page to show sequence diagrams etc, but IMO this is all pretty clear.

Can you recheck it?

hamishwillee committed:
Update files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md

hamishwillee reviewed (approved):
@chrisdavidmills I have rewritten the section on critical client hints to include what I consider key information. Done as a bunch of little suggestions which hopefully explain why I thought it was important.

I'm approving because I'd be happy to merge, but that section now needs a review from you because I'm not longer objective on it.

PS I really like client hints, and I think they are quite fun to document. Does this make me some kind of sicko? Really glad you're doing these.

hamishwillee commented:
Arrrggg, unrelated but I dont think https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints is in the sidebar. Where do you think it should go? Peer under https://developer.mozilla.org/en-US/docs/Web/HTTP/Session ?

bershanskiy commented on files/en-us/web/http/client_hints/index.md:
```diff
@@ -72,6 +72,33 @@ The high entropy hints are those that have the potential to give away more infor
+## Critical client hints
+
+A _critical client hint_ is one where applying the response may significantly change the rendered page, potentially in a way that is jarring or will significantly affect usability, and which must therefore be applied before the content is rendered.
+For example, `Sec-CH-Prefers-Reduced-Motion` is commonly treated as a critical hint, because it might markedly affect the behaviour of animations, and because a user who chooses this preference really needs for it to be set.
```
> `Sec-CH-Prefers-Reduced-Motion` is commonly treated as a critical hint

Interesting, is this observed in the wild? I had seen only a few sites using `Sec-CH-Prefers-Color-Scheme` to avoid ["flash of unstyled content"](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) typically flash of white content in dark color scheme) but I have never seen any issues related to animations.

hamishwillee replied:
No, this is creative writing. But it should be a critical hint - if a user sets a preference for reduced motion it is usually because they have an actual health risk. So this is my way of saying that it is optional, but a server that provides this kind of information should set it as critical.

chrisdavidmills committed:
copy edits for hamishwillee changes

chrisdavidmills commented:
@hamishwillee nice one, these changes are looking good! 

> Arrrggg, unrelated but I dont think https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints is in the sidebar. Where do you think it should go? Peer under https://developer.mozilla.org/en-US/docs/Web/HTTP/Session ?

Yes, that would make sense to me.

> PS I really like client hints, and I think they are quite fun to document. Does this make me some kind of sicko? Really glad you're doing these.

No more of a sicko that I thought previously anyway ;-)

chrisdavidmills committed:
empty commit to trigger job rerun

hamishwillee commented:
@chrisdavidmills Excellent. Did you see the comments against https://github.com/mdn/content/pull/22266 which covers much of the same ground - and I think there are some good cross links we could add?

Most of that could be done as a post process, but one comment is relevent now: https://github.com/mdn/content/pull/22266#discussion_r1022199652 ?
Specifically, because you are using Critical-CH I _think_ you must in the example also include `Vary` header, and if this is the case we should also mention this in the critical hints section. Can I leave it to you to confirm?

chrisdavidmills committed:
a couple more updates based on Hamish comments

chrisdavidmills commented:
> @chrisdavidmills Excellent. Did you see the comments against https://github.com/mdn/content/pull/22266 which covers much of the same ground - and I think there are some good cross links we could add?

I added the cross link to the SVG `animate` page from @bershanskiy 's PR — that was really useful.

> Most of that could be done as a post process, but one comment is relevent now: https://github.com/mdn/content/pull/22266#discussion_r1022199652 ?
Specifically, because you are using Critical-CH I think you must in the example also include Vary header, and if this is the case we should also mention this in the critical hints section. Can I leave it to you to confirm?

I've added a note to cover this, along with the header in the example. It was a real pain to track down this info, but I eventually found it in the `Critical-CH` spec — see https://datatracker.ietf.org/doc/html/draft-davidben-http-client-hint-reliability-00#section-3!

hamishwillee commented:
Looks good to me thanks @chrisdavidmills - let's get this in! 

Would be great if you can work with @bershanskiy on #22266

chrisdavidmills commented:
Nice one, thanks! I think for the other PR it would just make sense to follow the same structure as my page now it has been merged...happy to help tho @bershanskiy if you need anything.

bershanskiy commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -0,0 +1,96 @@
+HTTP/1.1 200 OK
+Content-Type: text/html
+Accept-CH: Sec-CH-Prefers-Reduced-Motion
+Vary: Sec-CH-Prefers-Color-Scheme
```
I'm not sure, but perhaps this was meant to be `Sec-CH-Prefers-Reduced-Motion` instead of `Sec-CH-Prefers-Color-Scheme`?

bershanskiy replied:
Specifically, line 70 says:
> We've also specified `Sec-CH-Prefers-Reduced-Motion` in the {{httpheader("Vary")}} header

chrisdavidmills replied:
@bershanskiy yup, you are right ;-)

I've submitted https://github.com/mdn/content/pull/22424 to fix it.
